### DISCUSSION TITLE
Sequence.Insert() feature request

### DISCUSSION THREAD

**Discussion by YZDwork** [KyryloKuzyk/PrimeTween#33] [Ideas]
1. Group + Delay cannot replace ChainCallback
	It seems that PrimeTween's sequence does not have time point control. Although we use ChainCallback and Group(Tween.Delay()) to set callbacks at the same time point, they have different behaviors.

2. It seems that we need "Insert"
	In your explanation, you mentioned that "Group" is the same as "Join" and we can use "Group(Tween)" instead of "Insert". However, this has raised a question. In DoTween, "Insert" does not impact the main Sequence, whereas "Join" is similar to "Append" and does affect the main Sequence. As the example shows, it's not always possible to replace "Insert" with "Group".
	Sometimes, we need callbacks to run during a Chain (when the Chain's duration is 1f, the callback needs to run at 0.5f). This means that ChainCallback cannot be used in this situation. Unfortunately, Group + Delay cannot work either.

I apologize for my poor English, and I may use some inappropriate words. Please forgive me.

Here is the project：[https://github.com/YZDwork/PrimeTweenDebug](https://github.com/YZDwork/PrimeTweenDebug)

**Comment by KyryloKuzyk** [comment]
> It seems that PrimeTween's sequence does not have time point control.

Since version 1.1.0 sequences do have a time control, and you can even manually set the elapsedTime on them.

> Although we use ChainCallback and Group(Tween.Delay()) to set callbacks at the same time point, they have different behaviors.

'Group' and 'Chain' are two completely different operations. 'Grouped' tweens start at the same time and play in **parallel** to one another. But when you use 'Chain', the chained tween will start playing only **after all** tweens before it.

> It seems that we need "Insert". Sometimes, we need callbacks to run during a Chain (when the Chain's duration is 1f, the callback needs to run at 0.5f). 

Yes, there is no Insert() operation in PrimeTween. It can be added very easily, although I don't see good usecases for it for now. For example, your use case can look like this:
```csharp
// console output:
// tween1 completed at 1.00
// tween2 completed at 2.00
// Callback at the middle of tween3 at time 3.50
// tween3 completed at 5.00
Sequence tween3WithCallbackAtMiddle = Sequence.Create(tween3)
    .Group(Tween.Delay(tween3.duration / 2f, () => print($"Callback at the middle of tween3 at time {Time.time:F2}")));

Sequence.Create()
    .Group(tween1) // tween1 and tween2 run in parallel
    .Group(tween2)
    .Chain(tween3WithCallbackAtMiddle); // tween3 starts after tween1 and tween2; the callback is invoked at the middle of tween3
```

Or the same as above, but simplified:
```csharp
Sequence.Create()
    .Group(tween1)
    .Group(tween2)
    .Chain(tween3.Group(Tween.Delay(tween3.duration / 2f, () => print($"Callback at the middle of tween3 at time {Time.time:F2}"))));
```

I'll think about adding Insert() in future version. Please share more examples of its usage if you have.

**Reply by KyryloKuzyk** [reply]
Here is another way to insert a callback in the middle of a chained tween:
```csharp
Sequence.Create()
    .Group(tween1)
    .Chain(Tween.PositionY(transform, 10f, duration: 1f)))
    .Group(Tween.Delay(0.5f, () => print($"Callback at the middle of chained Tween.PositionY(...) at time {Time.time}")));
```
Group() operation groups with the 'previous' tween in a Sequence, so the delay will wait for tween1 and will execute the callback in the middle of Tween.PositionY(...).

**Reply by KyryloKuzyk** [reply]
@YZDwork So I looked at the [PrimeTweenTest.cs](https://github.com/YZDwork/PrimeTweenDebug/blob/main/Assets/PrimeTweenTest.cs) script in your project. It doesn't work as intended because the Group() operation on [line 21](https://github.com/YZDwork/PrimeTweenDebug/blob/8855a84b1a0c1766263796b4269ea4688582060c/Assets/PrimeTweenTest.cs#L21) groups the tween with the **previous** tween in the sequence and not with the beginning of the sequence. The previous tween starts at time == 2f, so you don't get an expected result.

The correct way to write this sequence is the following:
```csharp
var seq = Sequence.Create()
    .Group(Sequence.Create()
        .Chain(Tween.LocalPosition(Player, new Vector3(0, Target.localPosition.y, 0), 1f))
        .Chain(Tween.LocalPosition(Player, new Vector3(Target.localPosition.x, 0, 0), 1f))
        .Chain(Tween.LocalPosition(Player, Target.localPosition, 1f)))
    .Group(Sequence.Create()
        .Chain(Tween.Scale(Player, Vector3.one * 3f, 1.5f))
        .Chain(Tween.Scale(Player, Vector3.one, 1.5f)));
```

The code above starts two nested sequences and runs them in parallel to one another with a Group() operation. I understand this approach may not be as simple as using an Insert() operation, so I'm considering adding the Insert() API in the next version.

**Reply by YZDwork** [reply]
The help you provided was greatly appreciated,I already find another way to resolve this probrom.

**Comment by KyryloKuzyk** [comment]
I added sequence.Insert() and sequence.InsertCallback() APIs in version [1.1.10](https://github.com/KyryloKuzyk/PrimeTween/blob/main/changelog.md). 

While it's possible to create sequences of any complexity using just Group() and Chain() operations, the new Insert() API makes it easier in some cases. 

For example, the Demo scene used this code previously to insert a sequence at a particular place in the `animateAllSequence`. This code prepends a delay to a sequence, then groups the delayed sequence to the `animateAllSequence`:
```csharp
float delay = 0f;
foreach (var animatable in animatables) {
    Sequence sequence = animatable.Animate(toEndValue);
    Sequence delayed = Tween.Delay(delay).Chain(sequence);
    animateAllSequence.Group(delayed); 
    delay += 0.6f;
}
```

The new version uses Insert() and makes the code much easier to write and understand:
```csharp
float delay = 0f;
foreach (var animatable in animatables) {
    Sequence sequence = animatable.Animate(toEndValue);
    animateAllSequence.Insert(delay, sequence);
    delay += 0.6f;
}
```

@YZDwork Huge thanks for the feature request!

**Reply by YZDwork** [reply]
I apologize for the delay in responding. Thank you very much! It has been truly helpful! I hope this program will gain popularity worldwide.
